### PR TITLE
Fix fallback viewer if pages are unavailable

### PR DESCRIPTION
**Pull request by mbarton**
Pages can be missing in a few situations:

- An old upload that we haven't reprocessed yet
- `PdfOcrExtractor` failed whilst processing the file

In these cases there was code to fallback to the old viewer but it wasn't working. This PR fixes it by taking advantage of the page state now being in Redux to move the decision up a level to `Viewer` rather than `PageViewer`.

It also fixes a server-side error where we did `minBy` on the empty list of pages. That now returns a 404. As `Viewer` is now so complex ironically the simplest fix was to convert that to an empty list of pages in the Redux state and then fallback in that case. This is not ideal and the whole `Viewer` is very much due a refactor. I've added comments inline explaining our ideas for this.

### CODE REVIEW

File: docs/cheat-sheet.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Cheat sheet
```
Comment (mbarton) [inline docs/cheat-sheet.md:1]: @joelochlann should we move this into the runbook? Thinking about it there's a strong argument to start a `docs` folder in this repo with explainers, cheat sheets and tips per service

File: frontend/src/js/components/viewer/PageViewer/PageViewer.tsx [hunk 2]
```diff
@@ -39,13 +30,6 @@ export default function PageViewer({ uri, q, fallbackRenderFn }: Props) {
         }
     }, [currentHighlightElement]);
 
-    useEffect(() => {
```
Comment (mbarton) [inline frontend/src/js/components/viewer/PageViewer/PageViewer.tsx:42]: The initial load is now done in `Viewer`

File: frontend/src/js/actions/pages/loadPages.ts [hunk 3]
```diff
@@ -3,7 +3,11 @@ import { fetchPages } from "../../components/viewer/PageViewer/pageViewerApi";
 import { GiantAction, PagesActionType } from "../../types/redux/GiantActions";
 import { GiantState } from "../../types/redux/GiantState";
 
-export function loadPages(pageApiUri: string, viewportTop: number, viewportBottom: number, q?: string): ThunkAction<void, GiantState, void, GiantAction> {
+export function loadPages(uri: string, q?: string, maybeViewportTop?: number, maybeViewportBottom?: number): ThunkAction<void, GiantState, void, GiantAction> {
```
Comment (mbarton) [inline frontend/src/js/actions/pages/loadPages.ts:6]: I re-arranged the parameters to make the defaulting cleaner in `Viewer`

File: frontend/src/js/components/viewer/Viewer.tsx [hunk 4]
```diff
@@ -195,6 +215,7 @@ class Viewer extends React.Component<Props, State> {
     componentWillUnmount() {
         document.title = "Giant";
         this.props.resetResource();
+        this.props.resetPages();
```
Comment (mbarton) [inline frontend/src/js/components/viewer/Viewer.tsx:218]: Without this the old page state hangs around and we can't properly calculate the fallback logic if you click back to Search and then into the Viewer again via a seach result

File: scripts/setup.sh [hunk 5]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
```
Comment (mbarton) [inline scripts/setup.sh:1]: I forgot to include this script when we opened the repo 🤦

**Comment by mbarton** [comment]
[Targeted fix for the `empty minBy` error when using the Page Viewer in PROD](https://trello.com/c/yq1YSTS9/1349-targeted-fix-for-the-empty-minby-error-when-using-the-page-viewer-in-prod)

File: docs/cheat-sheet.md (hunk 1 above)
Comment (joelochlann) [reply docs/cheat-sheet.md:1]: yeah anything here that's not in runbook already, let's move there

**Review by joelochlann — APPROVED** [review]
I think this looks merge-able! I'm assuming the race between pages & full resource request doesn't make anything too weird for the user. And that intermediate loading states don't look too crazy etc

Would be good to refactor the pages state a bit soon, along the lines we discussed. enum between

- first page request hasn't happened yet
- page request pending
- request came back successfully
- request came back with an error

**Pull request merged**